src/gpl/src/initialPlace.cpp

```diff
@@ -106,23 +142,157 @@ void InitialPlace::doBicgstabPlace() {
+      CUDA_ERROR(cudaMalloc((void**)&d_csrRowIndY, (m+1) * sizeof(int)));
+      CUSPARSE_ERROR(cusparseXcoo2csr(handleCusparseX, d_cooRowIndexX, nnz_x, m, d_csrRowIndX, CUSPARSE_INDEX_BASE_ZERO));
+      CUSPARSE_ERROR(cusparseXcoo2csr(handleCusparseY, d_cooRowIndexY, nnz_y, m, d_csrRowIndY, CUSPARSE_INDEX_BASE_ZERO));
+      CUSOLVER_ERROR(cusolverSpScsrlsvqr(handleCusolverX, m, nnz_x, descrX, d_cooValX, d_csrRowIndX, d_cooColIndexX, d_fixedInstForceVecX_, tol, reorder, d_instLocVecX, &singularity));
+      CUSOLVER_ERROR(cusolverSpScsrlsvqr(handleCusolverY, m, nnz_y, descrY, d_cooValY, d_csrRowIndY, d_cooColIndexY, d_fixedInstForceVecY_, tol, reorder, d_instLocVecY, &singularity));
+      
+      // Sync and Copy data to host
+      CUDA_ERROR(cudaMemcpyAsync(instLocVecX_.data(), d_instLocVecX, sizeof(float)*m, cudaMemcpyDeviceToHost, streamX));
+      CUDA_ERROR(cudaMemcpyAsync(instLocVecY_.data(), d_instLocVecY, sizeof(float)*m, cudaMemcpyDeviceToHost, streamY));
+
+      // Destroy what is not needed in both of device and host
+      CUDA_ERROR(cudaFree(d_cooColIndexX));
+      CUDA_ERROR(cudaFree(d_cooRowIndexX));
+      CUDA_ERROR(cudaFree(d_cooValX));
+      CUDA_ERROR(cudaFree(d_csrRowIndX));
+      CUDA_ERROR(cudaFree(d_instLocVecX));
+      CUDA_ERROR(cudaFree(d_fixedInstForceVecX_));
+      CUDA_ERROR(cudaFree(d_cooColIndexY));
+      CUDA_ERROR(cudaFree(d_cooRowIndexY));
+      CUDA_ERROR(cudaFree(d_cooValY));
+      CUDA_ERROR(cudaFree(d_csrRowIndY));
+      CUDA_ERROR(cudaFree(d_instLocVecY));
+      CUDA_ERROR(cudaFree(d_fixedInstForceVecY_));
+      CUSPARSE_ERROR(cusparseDestroyMatDescr( descrX ) );
+      CUSPARSE_ERROR(cusparseDestroy(handleCusparseX));
+      CUSOLVER_ERROR(cusolverSpDestroy(handleCusolverX));  
+      CUSPARSE_ERROR(cusparseDestroyMatDescr( descrY ) );
+      CUSPARSE_ERROR(cusparseDestroy(handleCusparseY));
+      CUSOLVER_ERROR(cusolverSpDestroy(handleCusolverY));   
+
+      // std::cout << "GPU X: " << instLocVecX_[0] << ", " << instLocVecX_[1] << ", " << instLocVecX_[200] << std:: endl;
+      // std::cout << "GPU Y: " << instLocVecY_[0] << ", " << instLocVecY_[1] << ", " << instLocVecY_[200] << std:: endl;
+         
+    }
+    else{
+      // BiCGSTAB solver for initial place
+      BiCGSTAB< SMatrix, IdentityPreconditioner > solver;
+      solver.setMaxIterations(ipVars_.maxSolverIter);
+      solver.compute(placeInstForceMatrixX_);
+      instLocVecX_ = solver.solveWithGuess(fixedInstForceVecX_, instLocVecX_);
+      errorX = solver.error();
+      
+      solver.compute(placeInstForceMatrixY_);
+      instLocVecY_ = solver.solveWithGuess(fixedInstForceVecY_, instLocVecY_);
+      errorY = solver.error();
 
-    solver.compute(placeInstForceMatrixY_);
-    instLocVecY_ = solver.solveWithGuess(fixedInstForceVecY_, instLocVecY_);
-    errorY = solver.error();
+      log_->report("[InitialPlace]  Iter: {} CG residual: {:0.8f} HPWL: {}",
+        i, max(errorX, errorY), pb_->hpwl());
```

Comment:
This looks like possibly useful status info, is it possible for us to log it using the output of the GPU routine so somebody could e.g. run with GPU and without GPU and compare the status output side-by-side in the logs?

Reply:
Ha You found it. It should be outside. I just didn't find a good function to record the errorX and errorY for the GPU. I will look for a good method to record it for GPU version.

Reply:
I use the relative residual to output error information, the same information as the CPU gave.